tests/src/test/spark330/scala/org/apache/spark/sql/RapidsTestsTrait.scala

```diff
@@ -0,0 +1,356 @@
+      input: MapData,
+      kt: DataType,
+      vt: DataType): (ArrayData, ArrayData) = {
+    val keyArray = input.keyArray().toArray[Any](kt)
+    val valueArray = input.valueArray().toArray[Any](vt)
+    val newMap = (keyArray.zip(valueArray)).toMap
+    val sortedMap = mutable.SortedMap(newMap.toSeq: _*)(TypeUtils.getInterpretedOrdering(kt))
+    (new GenericArrayData(sortedMap.keys.toArray), new GenericArrayData(sortedMap.values.toArray))
+  }
+
+  override protected def checkResult(
+      result: Any,
+      expected: Any,
+      exprDataType: DataType,
+      exprNullable: Boolean): Boolean = {
+    val dataType = UserDefinedType.sqlType(exprDataType)
+
+    // The result is null for a non-nullable expression
+    assert(result != null || exprNullable, "exprNullable should be true if result is null")
+    (result, expected) match {
+      case (result: Array[Byte], expected: Array[Byte]) =>
+        java.util.Arrays.equals(result, expected)
+      case (result: Double, expected: Spread[Double @unchecked]) =>
+        expected.asInstanceOf[Spread[Double]].isWithin(result)
+      case (result: InternalRow, expected: InternalRow) =>
+        val st = dataType.asInstanceOf[StructType]
+        assert(result.numFields == st.length && expected.numFields == st.length)
+        st.zipWithIndex.forall {
+          case (f, i) =>
+            checkResult(
+              result.get(i, f.dataType),
+              expected.get(i, f.dataType),
+              f.dataType,
+              f.nullable)
+        }
+      case (result: ArrayData, expected: ArrayData) =>
+        result.numElements == expected.numElements && {
+          val ArrayType(et, cn) = dataType.asInstanceOf[ArrayType]
+          var isSame = true
+          var i = 0
+          while (isSame && i < result.numElements) {
+            isSame = checkResult(result.get(i, et), expected.get(i, et), et, cn)
+            i += 1
+          }
+          isSame
+        }
+      case (result: MapData, expected: MapData) =>
+        val MapType(kt, vt, vcn) = dataType.asInstanceOf[MapType]
+        checkResult(
+          getSortedArrays(result, kt, vt)._1,
+          getSortedArrays(expected, kt, vt)._1,
+          ArrayType(kt, containsNull = false),
+          exprNullable = false) && checkResult(
+          getSortedArrays(result, kt, vt)._2,
+          getSortedArrays(expected, kt, vt)._2,
+          ArrayType(vt, vcn),
+          exprNullable = false)
+      case (result: Double, expected: Double) =>
+        if (
+          (isNaNOrInf(result) || isNaNOrInf(expected))
+          || (result == -0.0) || (expected == -0.0)
+        ) {
+          java.lang.Double.doubleToRawLongBits(result) ==
+            java.lang.Double.doubleToRawLongBits(expected)
+        } else {
+          Precision.equalsWithRelativeTolerance(result, expected, 0.00001d)
+        }
+      case (result: Float, expected: Float) =>
+        if (expected.isNaN) result.isNaN else expected == result
```

Comment:
Odd that this has no -0.0 checking and no tolerance yet Double has both.

Reply:
My guess is that vanilla Spark does not have many assersions on Floats or the Float compararison issues is not a problem. So by default comparing floats is still using the most strict way (bitwise compare)